File: typescript/src/webhooks.ts
import { createHmac, timingSafeEqual } from 'node:crypto'

/**
 * Default tolerance window (in seconds) between the `t=` timestamp in the
 * signature header and the current time. Any signed payload older than this
 * is rejected — protects against captured-replay attacks. Five minutes.
 */
export const DEFAULT_SIGNATURE_TOLERANCE_SECONDS = 300

export interface VerifyOptions {
  /** Raw, unmodified request body. Reading as `await req.text()` or equivalent. */
  rawBody: string
  /** Value of the `X-Rerout-Signature` header. */
  signatureHeader: string
  /** Endpoint signing secret (`whsec_…`) from the dashboard. */
  secret: string
  /** Window in seconds; defaults to 300. Set 0 to disable timestamp check. */
  toleranceSeconds?: number
  /**
   * Injectable clock for testing. Returns the current time in unix seconds.
   * Defaults to `Math.floor(Date.now() / 1000)`.
   */
  now?: () => number
}

/**
 * Verify a Rerout webhook signature.
 *
 * The header format is `t={unix},v1={hex_hmac}`. The HMAC is computed over
 * `${timestamp}.${rawBody}` with the endpoint signing secret as the key.
 *
 * Returns `true` only when:
 * - the header parses cleanly,
 * - the timestamp is within `toleranceSeconds` of `now()`,
 * - and the computed HMAC matches the supplied `v1` in constant time.
 *
 * @example
 * ```ts
 * import { verifyReroutSignature } from '@rerout/sdk'
 *
 * app.post('/webhooks/rerout', async (req, res) => {
 *   const raw = await readRawBody(req)
 *   const ok = verifyReroutSignature({
 *     rawBody: raw,
 *     signatureHeader: req.headers['x-rerout-signature'] as string,
 *     secret: process.env.REROUT_WEBHOOK_SECRET!,
 *   })
 *   if (!ok) return res.status(400).end('bad signature')
 *   // handle event…
 *   res.status(200).end()
 * })
 * ```
 */
export function verifyReroutSignature(opts: VerifyOptions): boolean {
  if (!opts.signatureHeader || !opts.secret || opts.rawBody === undefined) {
    return false
  }
  const parts = parseSignatureHeader(opts.signatureHeader)
  if (!parts) return false

  const tolerance = opts.toleranceSeconds ?? DEFAULT_SIGNATURE_TOLERANCE_SECONDS
  if (tolerance > 0) {
    const now = opts.now ? opts.now() : Math.floor(Date.now() / 1000)
    if (Math.abs(now - parts.timestamp) > tolerance) return false
  }

  const expectedHex = createHmac('sha256', opts.secret)
    .update(`${parts.timestamp}.${opts.rawBody}`)
    .digest('hex')

  const expected = safeFromHex(expectedHex)
  const actual = safeFromHex(parts.v1)
  if (!expected || !actual || expected.length !== actual.length) return false
  return timingSafeEqual(expected, actual)
}

interface ParsedSignature {
  timestamp: number
  v1: string
}

function parseSignatureHeader(header: string): ParsedSignature | null {
  let timestamp: number | undefined
  let v1: string | undefined
  for (const segment of header.split(',')) {
    const eq = segment.indexOf('=')
    if (eq <= 0) continue
    const key = segment.slice(0, eq).trim().toLowerCase()
    const value = segment.slice(eq + 1).trim()
    if (key === 't') {
      const parsed = Number.parseInt(value, 10)
      if (Number.isFinite(parsed) && parsed > 0) timestamp = parsed
    } else if (key === 'v1') {
      v1 = value
    }
  }
  if (timestamp === undefined || !v1) return null
  return { timestamp, v1 }
}

function safeFromHex(hex: string): Buffer | null {
  if (hex.length === 0 || hex.length % 2 !== 0) return null
  if (!/^[0-9a-fA-F]+$/.test(hex)) return null
  return Buffer.from(hex, 'hex')
}

File: rust/src/webhooks.rs
//! Webhook signature verification — HMAC-SHA256 with constant-time compare.
//!
//! Rerout signs every delivery as `t={unix_seconds},v1={hex_hmac_sha256}`
//! where the HMAC is computed over `"{timestamp}.{raw_body}"` keyed by the
//! endpoint signing secret. This module exposes a single free function that
//! returns `true` if and only if the signature is valid.

use std::time::{SystemTime, UNIX_EPOCH};

use hmac::{Hmac, Mac};
use sha2::Sha256;
use subtle::ConstantTimeEq;

type HmacSha256 = Hmac<Sha256>;

/// Default tolerance window (in seconds) between the `t=` timestamp and the
/// current time. Five minutes — protects against captured-replay attacks.
pub const DEFAULT_TOLERANCE_SECONDS: i64 = 300;

/// Verify a Rerout webhook signature.
///
/// # Parameters
///
/// - `raw_body` — the raw, unmodified request body, exactly as received.
/// - `signature_header` — the value of the `X-Rerout-Signature` header.
/// - `secret` — the endpoint signing secret (`whsec_…`).
/// - `tolerance_seconds` — window in seconds the timestamp may drift. Pass `0`
///   to disable the staleness check.
/// - `now` — optional override for the current unix timestamp. Useful for
///   tests; defaults to `SystemTime::now()`.
///
/// Returns `true` only when the header parses cleanly, the timestamp is within
/// `tolerance_seconds` of `now`, and the computed HMAC matches the supplied
/// `v1` in constant time. Returns `false` for every malformed input —
/// callers should treat `false` as "reject this delivery".
///
/// # Example
///
/// ```
/// use rerout::webhooks::{verify_signature, DEFAULT_TOLERANCE_SECONDS};
/// # use hmac::{Hmac, Mac};
/// # use sha2::Sha256;
/// # type HmacSha256 = Hmac<Sha256>;
/// let secret = "whsec_test";
/// let body = r#"{"event":"link.created"}"#;
/// let ts = 1_700_000_000_i64;
/// # let mut mac = HmacSha256::new_from_slice(secret.as_bytes()).unwrap();
/// # mac.update(format!("{ts}.{body}").as_bytes());
/// # let sig = hex::encode(mac.finalize().into_bytes());
/// let header = format!("t={ts},v1={sig}");
/// let ok = verify_signature(
///     body,
///     &header,
///     secret,
///     DEFAULT_TOLERANCE_SECONDS,
///     Some(ts), // pretend "now" is the signing moment
/// );
/// assert!(ok);
/// ```
pub fn verify_signature(
    raw_body: &str,
    signature_header: &str,
    secret: &str,
    tolerance_seconds: i64,
    now: Option<i64>,
) -> bool {
    if signature_header.is_empty() || secret.is_empty() {
        return false;
    }

    let parsed = match parse_signature_header(signature_header) {
        Some(parts) => parts,
        None => return false,
    };

    if tolerance_seconds > 0 {
        let now_seconds = now.unwrap_or_else(current_unix_seconds);
        let drift = now_seconds.saturating_sub(parsed.timestamp).abs();
        if drift > tolerance_seconds {
            return false;
        }
    }

    let mut mac = match HmacSha256::new_from_slice(secret.as_bytes()) {
        Ok(m) => m,
        Err(_) => return false,
    };
    let signed_payload = format!("{}.{}", parsed.timestamp, raw_body);
    mac.update(signed_payload.as_bytes());
    let expected = mac.finalize().into_bytes();

    let actual = match hex_decode(&parsed.v1) {
        Some(bytes) => bytes,
        None => return false,
    };
    if expected.len() != actual.len() {
        return false;
    }
    expected.ct_eq(&actual).into()
}

struct ParsedSignature {
    timestamp: i64,
    v1: String,
}

fn parse_signature_header(header: &str) -> Option<ParsedSignature> {
    let mut timestamp: Option<i64> = None;
    let mut v1: Option<String> = None;
    for segment in header.split(',') {
        let Some(eq) = segment.find('=') else {
            continue;
        };
        if eq == 0 {
            continue;
        }
        let key = segment[..eq].trim().to_ascii_lowercase();
        let value = segment[eq + 1..].trim();
        if value.is_empty() {
            continue;
        }
        match key.as_str() {
            "t" => {
                if let Ok(parsed) = value.parse::<i64>()
                    && parsed > 0
                {
                    timestamp = Some(parsed);
                }
            }
            "v1" => {
                v1 = Some(value.to_string());
            }
            _ => {}
        }
    }
    Some(ParsedSignature {
        timestamp: timestamp?,
        v1: v1?,
    })
}

fn hex_decode(s: &str) -> Option<Vec<u8>> {
    if s.is_empty() || s.len() % 2 != 0 {
        return None;
    }
    if !s.bytes().all(|b| b.is_ascii_hexdigit()) {
        return None;
    }
    hex::decode(s).ok()
}

fn current_unix_seconds() -> i64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_secs() as i64)
        .unwrap_or(0)
}
